Mission 1: 프로젝트 세팅 및 createElement 구현

youngseo-kangg opened the pull request:
## week1) 프로젝트 세팅 및 createElement 구현

### 프로젝트 세팅
- vite, typescript, babel을 통해 세팅했습니다.
- prettier 관련해서는 .prettierrc를 작성하여 적용하였고, eslint 관련해서는 rushstack 패키지를 적용해 보았습니다. 
  - 코드 내용이 얼마나 방대해 질 지는 모르겠지만, 일단 더 작성하기 전에 초반부터 적용하는게 개인적으로는 편해서 적용해 보았습니다.
  - rushstck이 eslint-config-airbnb보다 조금 더 유연하고 규칙의 적용 근거가 설명이 잘 되어있다고 느껴서 선택해보았습니다.
- 프로젝트에 필요한 패키지 설치를 위한 `yarn install`를 실행하고, README.md를 참고하여 필요한 script를 실행하면 관련된 내용을 볼 수 있습니다.

### createElement 구현
- typescript로 React.createElement를 만들어 보았습니다.
- 개인적으로는 createElement 함수의 매개변수 중 type에 대해 더 알아봐야 할 것 같습니다.
```
createElement: (type, props, ...children) => {
    return typeof type === "function"
      ? { name: type.name, props, children: type({ ...props, children }) }
      : {
          name: type,
          props,
          children: children.filter((child) => !!child)
        };
  }
```
- `typeof`를 사용해보니 중첩된 요소가 있는 경우에는 `type`이 `function`으로 잡히는 것 같은데, .name이라는 프로퍼티도 있으면서 props와 children을 받는 함수라고 예측만 되서.. 정확히 어떤 모습인지가 궁금한데, 어떻게 뜯어봐야할지가 애매한 것 같습니다.

youngseo-kangg committed:
feat: 세팅 및 ts(x) -> js(x) 트랜스파일 관련 환경 구축 및 확인
- 기본 README.md 작성
- babel, eslint, prettier, typescript 세팅 및 관련 config 작성
- ts(x) -> js(x) 트랜스파일 결과 확인용 컴포넌트 작성

youngseo-kangg committed:
chore: ESLint, Babel 설정 수정
- eslint rule 사용 관련 require(rushstack/eslint-patch) 추가
- 트랜스파일 시 주석 제거 관련 comments: false 옵션 추가

youngseo-kangg committed:
fix: eslint ignorePatterns typo 수정

youngseo-kangg committed:
feat: createElement 함수 및 관련 타입 작성, 결과 확인

youngseo-kangg committed:
feat: 컴포넌트 작성 및 html 속성 적용,확인

youngseo-kangg committed:
rushstack 관련 README.md 내용 제거

- 잘못된 내용 제거

YOOJS1205 commented on babel.config.js:
Babel로 직접 설정해주셨군요!

`vite.config.ts`의 `esbuild` 옵션으로도 쉽게 설정이 가능합니다. 확인해보시고 적용해보세요!

YOOJS1205 commented on tsconfig.json:
```diff
@@ -12,6 +12,8 @@
     "isolatedModules": true,
     "moduleDetection": "force",
     "noEmit": true,
+    "jsx": "preserve", // JSX 코드 변환 방식 설정
```
`preserve` 옵션을 사용하신 이유는 무엇인가요?
자동으로 `createElement`를 호출하여 출력하는 옵션이 있어요! 해당 옵션을 사용하면 JSX 구문을 자동으로 JavaScript 객체로 반환해줍니다. 연결된 링크 확인해주세요!

https://www.typescriptlang.org/ko/docs/handbook/jsx.html

YOOJS1205 commented on src/react/index.ts:
```diff
@@ -0,0 +1,64 @@
+  ) => IReactNode;
+}
+const React: IReact = {
+  createElement: (type, props: ElementPropsType, ...children: ChildrenType) => {
```
`createElement` 함수 구현의 흐름은 전반적으로 잘 해주셨습니다! 몇 가지 개선해야할 점들을 가이드 해드릴게요!

* `props`는 없는 경우도 있기 때문에, `ElementPropsType | null` 타입이 더 확장성 있을 것 같습니다.
* 자식 요소들에 평탄화 (flatten) 처리가 필요합니다!

  현재 createElement 함수를 기반으로 아래의 요소 렌더링을 시도해보세요!

  ```tsx
  const App = () => (
    <div>
      {[1, 2, 3].map((num) => (
        <span>{num}</span>
      ))}
    </div>
  );
  ```
  JSX에서 반환된 JavaScript 객체가 중첩 구조일 때 평탄화를 통해서 DOM API가 처리할 수 있도록 해야 합니다. 해당 내용 관련 리서치 해보시고, 반영하면 좋을 것 같아요!

YOOJS1205 commented on src/react/index.ts:
```diff
@@ -0,0 +1,64 @@
+/**
+ * React 스타일의 가상 DOM 노드를 나타냅니다.
+ */
+interface IReactNode {
```
IReactNode에서 개선되어야할 점들이 보이네요!

* `name`에는 HTML tag 명이 들어가게 될텐데, 아래의 타입이 더 명확하지 않을까요?!

  ```ts
  name: keyof HTMLElementTagNameMap;
  ```

* `props`는 없는 경우도 있기 때문에 `null`도 받을 수 있으면 좋을 것 같아요!
  ```ts
  props: ElementPropsType | null;
  ```

* `children`은 optional하게 바꿔주면 더 간결하게 작성이 가능할 것 같네요!
  ```ts
  children?: ChildrenType;
  ```

YOOJS1205 reviewed (commented):
고생 많으셨습니다!

구현된 `createElement`를 자동으로 호출하여 JSX 구문이 JavaScript 객체로 변환되는 트랜스파일링 세팅이 필요할 것 같아요!

코멘트 확인해보시고 개선되면 좋을 것 같습니다.

youngseo-kangg committed:
fix: babel 관련 내용 및 구조 수정, react 타입 관련 수정
- babel.config.js에 plugin 제거 및 vite.config.js esbuild 옵션 설정
- lib 내부로 react 관련 파일 이동
- react 관련 path '@' 관련 설정 추가
- component 내부로 커스텀 컴포넌트 관련 파일 이동
